File: src/ios/Agent/Chat/PlanGate.swift
import Foundation

/// Plan-confirmation gate for 深度龙虾Ai (Deep Agent Mode) — Layer B.
///
/// Unlike Layer A (persistent rules + proactive memory), the plan gate adds a
/// *hard* interaction: when deep mode is on and the agent's first reply to a
/// multi-step task is a plan-only turn (no tool calls, plan wrapped in the
/// fenced `plan` block the deep-mode fragment instructs it to emit), the UI
/// raises a confirm/edit bar instead of silently proceeding. The user then
/// either approves (which re-sends an "execute the plan" prompt through the
/// normal `send()` path) or edits/cancels.
///
/// The gate is deliberately fail-safe: if the model ignores the instruction
/// and starts acting immediately, `detectPlan` returns `nil`, the state stays
/// `.idle`, and nothing changes — a missed gate degrades to the pre-existing
/// behavior, never to a stuck or blocked thread.
enum PlanGate {

    enum State: Equatable {
        case idle
        case awaitingApproval(planText: String)
    }

    /// Fenced-block marker the deep-mode fragment tells the model to wrap its
    /// plan in. Detection matches the raw markdown before rendering.
    static let planMarker = "```plan"

    /// True when the assistant turn produced only text/thinking blocks and no
    /// tool-call blocks — i.e. it planned instead of acting.
    static func turnHasNoToolCalls(_ msg: ChatMessage) -> Bool {
        !msg.blocks.contains { block in
            switch block.kind {
            case .text, .thinking:
                return false
            default:
                return true
            }
        }
    }

    /// Combined text of all text blocks in a turn.
    static func visibleText(_ msg: ChatMessage) -> String {
        msg.blocks.compactMap { block -> String? in
            if case .text = block.kind { return block.content }
            return nil
        }.joined(separator: "\n")
    }

    /// Detect a plan-only turn. Returns the plan text when the turn has no
    /// tool calls AND its visible text contains the plan marker; nil otherwise.
    static func detectPlan(in msg: ChatMessage) -> String? {
        guard turnHasNoToolCalls(msg) else { return nil }
        let text = visibleText(msg)
        guard !text.isEmpty else { return nil }
        guard text.localizedCaseInsensitiveContains(planMarker) else { return nil }
        return text
    }
}

File: src/android/app/src/main/java/com/openminis/app/agent/PlanGate.kt
package com.openminis.app.agent

/**
 * Plan-confirmation gate for 深度龙虾Ai (Deep Agent Mode) — Layer B.
 *
 * When deep mode is on and the agent's first reply to a multi-step task is a
 * plan-only turn (no tool calls, plan wrapped in the fenced `plan` block the
 * deep-mode fragment instructs it to emit), the UI raises a confirm/edit bar
 * instead of silently proceeding.
 *
 * Fail-safe: if the model ignores the instruction and starts acting immediately,
 * `detectPlan` returns null and the state stays idle — a missed gate degrades to
 * the pre-existing "start right away" behavior, never to a stuck state.
 */
object PlanGate {

    enum class State {
        IDLE,
        AWAITING_APPROVAL,
    }

    /** Fenced-block marker the deep-mode fragment tells the model to wrap its plan in. */
    const val PLAN_MARKER = "```plan"

    /**
     * Detect a plan-only turn. Returns the plan text when the turn has no
     * tool calls AND its visible text contains the plan marker; null otherwise.
     */
    fun detectPlan(text: String, hasToolCalls: Boolean): String? {
        if (hasToolCalls) return null
        if (text.isBlank()) return null
        if (!text.contains(PLAN_MARKER, ignoreCase = true)) return null
        return text
    }

    /**
     * System prompt fragment explaining the planning phase.
     * Injected only when deep mode is enabled.
     */
    val systemPromptFragment: String
        get() = """
PLANNING PHASE — on the FIRST turn of a multi-step task, BEFORE calling any tools, output your plan wrapped in a fenced code block:

\`\`\`plan
1. <step 1>
2. <step 2>
...
\`\`\`

Then STOP — do not call any tools on this turn. The client will show the plan to the user for approval. Execution begins only after the user approves.

Rules:
- Only emit a plan on the FIRST turn of a new task. On follow-up turns, go straight to execution.
- The plan block must be the main content of the turn — no tool calls, just the plan and brief context.
- Be specific: list concrete steps, not vague intentions.
- For simple one-step tasks (e.g. "read this file"), skip the plan and go straight to execution.
"""
}
